packages/validator/src/services/prepareBeaconProposer.ts

```diff
@@ -1,47 +1,71 @@
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {fromHexString} from "@chainsafe/ssz";
+import {SLOTS_PER_EPOCH} from "@chainsafe/lodestar-params";
 import {IClock, ILoggerVc} from "../util/index.js";
 import {Metrics} from "../metrics.js";
 import {ValidatorStore} from "./validatorStore.js";
 import {IndicesService} from "./indices.js";
 
+const REGISTRATION_CHUNK_SIZE = 512;
 /**
  * This service is responsible for updating the BNs and/or Mev relays with
  * the corresponding feeRecipient suggestion. This should ideally run per epoch
  * but can be run per slot. Lighthouse also uses this to trigger any block
  */
 export class PrepareBeaconProposerService {
   constructor(
+    private readonly config: IBeaconConfig,
     private readonly logger: ILoggerVc,
     private readonly api: Api,
     private clock: IClock,
     private readonly validatorStore: ValidatorStore,
     private readonly indicesService: IndicesService,
-    private readonly metrics: Metrics | null
+    private readonly metrics: Metrics | null,
+    builderOpts: {enabled?: boolean}
   ) {
+    const currentEpoch = clock.getCurrentEpoch();
+
     clock.runEveryEpoch(this.prepareBeaconProposer);
+    void this.prepareBeaconProposer(currentEpoch);
+
+    if (builderOpts?.enabled) {
+      clock.runEveryEpoch(this.registerValidator);
+      void this.registerValidator(currentEpoch);
+    }
   }
 
   private prepareBeaconProposer = async (epoch: Epoch): Promise<void> => {
+    // Before bellatrix we don't need to update this data on bn/builder
+    if (epoch < this.config.BELLATRIX_FORK_EPOCH - 1) return;
+
     await Promise.all([
       // Run prepareBeaconProposer immediately for all known local indices
-      this.api.validator
-        .prepareBeaconProposer(this.getProposerData(this.indicesService.getAllLocalIndices()))
-        .catch((e: Error) => {
+      this.prepareBeaconProposerInChunks(this.getProposerData(this.indicesService.getAllLocalIndices())).catch(
+        (e: Error) => {
           this.logger.error("Error on prepareBeaconProposer", {epoch}, e);
-        }),
+        }
+      ),
 
       // At the same time fetch any remaining unknown validator indices, then poll duties for those newIndices only
       this.indicesService
         .pollValidatorIndices()
-        .then((newIndices) => this.api.validator.prepareBeaconProposer(this.getProposerData(newIndices)))
+        .then((newIndices) => this.prepareBeaconProposerInChunks(this.getProposerData(newIndices)))
         .catch((e: Error) => {
           this.logger.error("Error on poll indices and prepareBeaconProposer", {epoch}, e);
         }),
     ]);
   };
 
+  private prepareBeaconProposerInChunks = async (
+    proposers: routes.validator.ProposerPreparationData[]
+  ): Promise<void> => {
+    const proposerChunks = spliceIntoChunks(proposers, REGISTRATION_CHUNK_SIZE);
+    await Promise.all(proposerChunks.map((proposerChunk) => this.api.validator.prepareBeaconProposer(proposerChunk)));
```

Comment:
Would it better to not spam the beacon node with all chunks at once, but do this sequentially?

Reply:
sure makes sense :+1:

Reply:
done